File: go/workflows/account_transfer_workflow_scenarios.go
package workflows

import (
	"fmt"
	"money-transfer-worker/activities"
	"money-transfer-worker/app"
	"money-transfer-worker/messages"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

const (
	ADVANCED_VISIBILTY = "AccountTransferWorkflowAdvancedVisibility"
	NEEDS_APPROVAL     = "AccountTransferWorkflowHumanInLoop"
	BUG                = "AccountTransferWorkflowRecoverableFailure"
	APPROVAL_TIME      = 30
)

var stepKey = temporal.NewSearchAttributeKeyKeyword("Step")

func AccountTransferWorkflowScenarios(ctx workflow.Context, input app.TransferInput) (output *app.TransferOutput, err error) {
	name := workflow.GetInfo(ctx).WorkflowType.Name
	logger := workflow.GetLogger(ctx)
	logger.Info("Dynamic Account Transfer workflow started", "name", name)

	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    1 * time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	// Expose transfer status as a query
	ts, err := messages.SetQueryHandlerForStatus(ctx)
	if err != nil {
		return nil, err
	}

	var idempotencyKey string
	_ = workflow.SideEffect(ctx, func(ctx workflow.Context) interface{} {
		return uuid.New().String()
	}).Get(&idempotencyKey)

	// Validate
	upsertStep(ctx, "Validate")
	err = workflow.ExecuteActivity(ctx, activities.Validate, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 1, ts, 25)

	if NEEDS_APPROVAL == name {
		logger.Info("Waiting on 'approveTransfer' Signal or Update for workflow ID: ", workflow.GetInfo(ctx).WorkflowExecution.ID)
		updateProgress(ctx, 0, ts, 30)
		ts.TransferState = "waiting"

		// Wait for the approval for up to approvalTime
		c := messages.GetSignalChannelForApproval(ctx)
		receivedApproval, _ := c.ReceiveWithTimeout(ctx, time.Second*APPROVAL_TIME, nil)

		// If the signal was not received within the timeout, fail the workflow
		if !receivedApproval {
			return nil, fmt.Errorf("approval not received within %d seconds", APPROVAL_TIME)
		}
	}

	// Withdraw
	upsertStep(ctx, "Withdraw")
	err = workflow.ExecuteActivity(ctx, activities.Withdraw, idempotencyKey, input.Amount, name).Get(ctx, nil)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 3, ts, 50)

	if BUG == name {
		// Simulate bug
		panic("Simulated bug - fix me!")
	}

	// Deposit
	upsertStep(ctx, "Deposit")
	depositResponse := app.DepositResponse{}
	err = workflow.ExecuteActivity(ctx, activities.Deposit, idempotencyKey, input.Amount, name).Get(ctx, &depositResponse)
	if err != nil {
		// if deposit fails in an unrecoverable way, rollback the withdrawal and fail the workflow
		logger.Info("Deposit failed unrecoverable error, reverting withdraw")

		// Undo Withdraw (rollback)
		workflow.ExecuteActivity(ctx, activities.UndoWithdraw, input.Amount).Get(ctx, nil)

		return nil, fmt.Errorf("deposit failed %w", err)
	}
	updateProgress(ctx, 1, ts, 75)

	// Send Notification
	upsertStep(ctx, "SendNotification")
	err = workflow.ExecuteActivity(ctx, activities.SendNotification, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}
	updateProgress(ctx, 1, ts, 100)

	output = &app.TransferOutput{
		DepositResponse: depositResponse,
	}
	return output, nil
}

func upsertStep(ctx workflow.Context, step string) {
	if ADVANCED_VISIBILTY == workflow.GetInfo(ctx).WorkflowType.Name {
		workflow.GetLogger(ctx).Info("Advanced visibility ..", "step", step)
		workflow.UpsertTypedSearchAttributes(ctx, stepKey.ValueSet(step))
	}
}

File: ruby/workflows/account_transfer_workflow_scenarios.rb
# frozen_string_literal: true

require_relative 'account_transfer_workflow'
require 'temporalio/error'
require 'timeout'

module Workflows
  class AccountTransferWorkflowScenarios < AccountTransferWorkflow
    # This workflow is designed to demonstrate various scenarios for the Account Transfer Workflow.

    # `workflow_dynamic` makes this workflow to catch all the workflow types that is not otherwise
    # registered by the worker, i.e. AccountTransferWorkflowRecoverableFailure,
    # AccountTransferWorkflowHumanInLoop and AccountTransferWorkflowAdvancedVisibility.
    workflow_dynamic

    BUG = 'AccountTransferWorkflowRecoverableFailure'
    NEEDS_APPROVAL = 'AccountTransferWorkflowHumanInLoop'
    ADVANCED_VISIBILITY = 'AccountTransferWorkflowAdvancedVisibility'

    APPROVAL_TIME = 30

    attr_reader :approved

    def initialize
      super
      @approved = false
    end

    def execute(input)
      workflow_type = Temporalio::Workflow.info.workflow_type
      logger.info("Dynamic Account Transfer Workflow started: #{workflow_type}, input: #{input.to_h}")

      idempotency_key = Temporalio::Workflow.random.uuid

      upsert_step('Validate')
      validate_transfer(input)
      update_progress(25, 1)

      handle_approval_if_needed(workflow_type)

      upsert_step('Withdraw')
      withdraw_funds(idempotency_key, input.amount, workflow_type)
      update_progress(50, 3)

      handle_bug_simulation(workflow_type)

      upsert_step('Deposit')
      begin
        @deposit_response = deposit_funds(idempotency_key, input.amount, workflow_type)
        update_progress(75, 1)
      rescue StandardError => e
        logger.info('Deposit failed unrecoverable error, reverting withdraw')
        undo_withdraw(input.amount)
        raise StandardError, "Deposit failed: #{e.message}"
      end

      upsert_step('SendNotification')
      send_notification(input)
      update_progress(100, 1, 'finished')

      Models::TransferOutput.new(deposit_response: @deposit_response).deep_camelize_keys
    end

    workflow_signal(name: 'approveTransfer')
    def approve_transfer_signal
      logger.info('Approve Signal Received')
      if @transfer_state == 'waiting'
        @approved = true
      else
        logger.info('Approval not applied. Transfer is not waiting for approval')
      end
    end

    workflow_query(name: 'transferStatus')
    def query_transfer_status
      Models::TransferStatus.new(
        progress_percentage: progress,
        transfer_state: transfer_state,
        workflow_status: '',
        charge_result: deposit_response,
        approval_time: APPROVAL_TIME
      ).deep_camelize_keys
    end

    private

    def handle_approval_if_needed(workflow_type)
      return unless workflow_type == NEEDS_APPROVAL

      logger.info("Waiting on 'approveTransfer' Signal for workflow ID: #{Temporalio::Workflow.info.workflow_id}")
      update_progress(30, 0, 'waiting')

      begin
        Temporalio::Workflow.timeout(APPROVAL_TIME) { Temporalio::Workflow.wait_condition { approved } }
      rescue Timeout::Error
        logger.error("Approval not received within #{APPROVAL_TIME} seconds")
        raise Temporalio::Error::ApplicationError.new("Approval not received within #{APPROVAL_TIME} seconds",
                                                      non_retryable: true)
      end
    end

    def handle_bug_simulation(workflow_type)
      return unless workflow_type == BUG

      raise 'Simulated bug - fix me!'
    end

    def undo_withdraw(amount)
      Temporalio::Workflow.execute_activity(
        'Activities::UndoWithdrawActivity',
        amount,
        start_to_close_timeout: 5,
        retry_policy: Activities::BaseActivity::RETRY_POLICY
      )
    end

    def upsert_step(step)
      return unless Temporalio::Workflow.info.workflow_type == ADVANCED_VISIBILITY

      logger.info("Advanced visibility... On step: #{step}")
      Temporalio::Workflow.upsert_search_attributes('Step' => step)
    end
  end
end
